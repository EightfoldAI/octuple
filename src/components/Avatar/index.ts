export * from './Avatar.types';
export * from './Avatar';
