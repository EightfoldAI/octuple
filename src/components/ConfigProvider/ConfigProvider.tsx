'use client';

import React, {
  createContext,
  FC,
  useContext,
  useEffect,
  useState,
} from 'react';
import LocaleReceiver from '../LocaleProvider/LocaleReceiver';
import LocaleProvider from '../LocaleProvider';
import { registerFont, registerTheme } from './Theming/styleGenerator';
import { ConfigProviderProps, IConfigContext } from './ConfigProvider.types';
import {
  IRegisterFont,
  IRegisterTheme,
  FontOptions,
  ThemeOptions,
} from './Theming';
import { useFocusVisibleClassName } from '../../hooks/useFocusVisibleClassName';
import { ParentComponentsContextProvider } from './ParentComponentsContext';
import { DisabledContextProvider } from './DisabledContext';
import { GradientContextProvider } from './GradientContext';
import { ShapeContextProvider } from './ShapeContext';
import { SizeContextProvider } from './SizeContext';
import { ValidateMessages } from '../Form/Internal/OcForm.types';
import { OcFormProvider } from '../Form/Internal';
import defaultLocale from '../Locale/Default';
import { canUseDocElement } from '../../shared/utilities';

const ConfigContext: React.Context<Partial<IConfigContext>> = createContext<
  Partial<IConfigContext>
>({});

const DEFAULT_THEME: string = 'blue';

const DEFAULT_FOCUS_VISIBLE: boolean = true;
const DEFAULT_FOCUS_VISIBLE_ELEMENT: HTMLElement = canUseDocElement()
  ? document.documentElement
  : null;

const ConfigProvider: FC<ConfigProviderProps> = ({
  children,
  componentName,
  disabled = false,
  focusVisibleOptions = {
    focusVisible: DEFAULT_FOCUS_VISIBLE,
    focusVisibleElement: DEFAULT_FOCUS_VISIBLE_ELEMENT,
  },
  fontOptions: defaultFontOptions,
  form,
  gradient = false,
  icomoonIconSet = {},
  locale,
  shape,
  size,
  themeOptions: defaultThemeOptions,
}) => {
  const [fontOptions, setFontOptions] =
    useState<FontOptions>(defaultFontOptions);

  const [registeredFont, setRegisteredFont] = useState<IRegisterFont>(
    {} as IRegisterFont
  );

  const [themeOptions, setThemeOptions] =
    useState<ThemeOptions>(defaultThemeOptions);

  const [registeredTheme, setRegisteredTheme] = useState<IRegisterTheme>(
    {} as IRegisterTheme
  );

  if (focusVisibleOptions?.focusVisible) {
    useFocusVisibleClassName(
      focusVisibleOptions?.focusVisible,
      focusVisibleOptions?.focusVisibleElement
    );
  }

  useEffect(() => {
    if (themeOptions) {
      setRegisteredTheme(
        registerTheme({
          name: DEFAULT_THEME,
          ...themeOptions,
        })
      );
    }
  }, [themeOptions]);

  useEffect(() => {
    if (fontOptions) {
      setRegisteredFont(
        registerFont({
          ...fontOptions,
        })
      );
    }
  }, [fontOptions]);

  let childNode = children;
  let validateMessages: ValidateMessages = {};

  if (locale) {
    validateMessages =
      locale.Form?.defaultValidateMessages ||
      defaultLocale.Form?.defaultValidateMessages ||
      {};
  }

  if (form?.validateMessages) {
    validateMessages = { ...validateMessages, ...form.validateMessages };
  }

  if (Object.keys(validateMessages).length > 0) {
    childNode = (
      <OcFormProvider validateMessages={validateMessages}>
        {children}
      </OcFormProvider>
    );
  }

  if (locale) {
    childNode = <LocaleProvider locale={locale}>{childNode}</LocaleProvider>;
  }

  if (shape) {
    childNode = (
      <ShapeContextProvider shape={shape}>{childNode}</ShapeContextProvider>
    );
  }

  if (size) {
    childNode = (
      <SizeContextProvider size={size}>{childNode}</SizeContextProvider>
    );
  }

  if (gradient !== undefined) {
    childNode = (
      <GradientContextProvider gradient={gradient}>
        {childNode}
      </GradientContextProvider>
    );
  }

  if (disabled !== undefined) {
    childNode = (
      <DisabledContextProvider disabled={disabled}>
        {childNode}
      </DisabledContextProvider>
    );
  }

  if (componentName !== undefined) {
    childNode = (
      <ParentComponentsContextProvider componentName={componentName}>
        {childNode}
      </ParentComponentsContextProvider>
    );
  }

  return (
    <LocaleReceiver>
      {(_, __) => (
        <ConfigContext.Provider
          value={{
            fontOptions,
            themeOptions,
            setFontOptions,
            setThemeOptions,
            disabled,
            focusVisibleOptions,
            form,
            icomoonIconSet,
            locale,
            registeredFont,
            registeredTheme,
            shape,
            size,
          }}
        >
          {childNode}
        </ConfigContext.Provider>
      )}
    </LocaleReceiver>
  );
};

const useConfig = () => {
  const context = useContext(ConfigContext);
  if (context === undefined) {
    throw new Error('useConfig hook must be used within config provider');
  }
  return context;
};

export { ConfigProvider, useConfig };
