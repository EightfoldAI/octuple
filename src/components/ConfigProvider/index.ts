export * from './ConfigProvider.types';
export * from './ConfigProvider';
export * from './Theming';
