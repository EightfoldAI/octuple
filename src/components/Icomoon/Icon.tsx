import React, { FC, useContext } from 'react';
import { mergeClasses } from '../../shared/utilities';
import { IcomoonProps, IcomoonSize } from './';
import IcomoonReact from 'icomoon-react';

import styles from './icon.module.scss';

const IconSetContext = React.createContext<object | null>(null);

export const IcomoonProvider = IconSetContext.Provider;

export const Icomoon: FC<IcomoonProps> = ({
    ariaHidden = false,
    classNames,
    color,
    iconName,
    id,
    role = 'presentation',
    size = IcomoonSize.Medium,
    'data-test-id': dataTestId,
}) => {
    const iconClassNames: string = mergeClasses([
        classNames,
        styles.iconWrapper,
    ]);

    return (
        <span
            data-test-id={dataTestId}
            aria-hidden={ariaHidden}
            className={iconClassNames}
            id={id}
            role={role}
        >
            <IconSetContext.Consumer>
                {(iconSet) => (
                    <IcomoonReact
                        iconSet={iconSet}
                        icon={iconName}
                        size={size}
                        color={color}
                    />
                )}
            </IconSetContext.Consumer>
        </span>
    );
};
