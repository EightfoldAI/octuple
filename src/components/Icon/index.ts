export * from './Icon';
export * from './Icon.types';
export * from './mdi';
