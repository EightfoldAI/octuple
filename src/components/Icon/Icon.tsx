'use client';

import React, { FC } from 'react';
import IcomoonReact from 'icomoon-react';

import { mergeClasses } from '../../shared/utilities';
import { Icon as MdiIcon } from '@mdi/react';
import { IconProps, IconSize } from './index';

import styles from './icon.module.scss';
import { useConfig } from '../ConfigProvider';

export const Icon: FC<IconProps> = ({
  ariaHidden = false,
  classNames,
  color,
  description,
  horizontal,
  id,
  path,
  role = 'presentation',
  rotate,
  size = IconSize.Medium,
  spin,
  style,
  title,
  vertical,
  'data-test-id': dataTestId,
  icomoonIconName,
}) => {
  const { icomoonIconSet } = useConfig();

  const iconClassNames: string = mergeClasses([classNames, styles.iconWrapper]);

  if (!icomoonIconName && !path) {
    return null;
  }

  const iconComponent = icomoonIconName ? (
    <IcomoonReact
      iconSet={icomoonIconSet}
      size={size}
      color={color}
      icon={icomoonIconName}
      role="presentation"
    />
  ) : (
    <MdiIcon
      color={color}
      description={description}
      horizontal={horizontal}
      path={path}
      rotate={rotate}
      size={size}
      title={title}
      vertical={vertical}
      spin={spin}
    />
  );

  return (
    <span
      data-test-id={dataTestId}
      aria-hidden={ariaHidden}
      className={iconClassNames}
      id={id}
      role={role}
      style={style ? style : null}
    >
      {iconComponent}
    </span>
  );
};

Icon.displayName = 'Icon';
