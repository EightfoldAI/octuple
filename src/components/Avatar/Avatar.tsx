'use client';

import React, { FC, Ref, useEffect, useMemo, useRef, useState } from 'react';
import {
  AvatarFallbackProps,
  AvatarIconProps,
  AvatarOutlineProps,
  AvatarProps,
  BaseAvatarProps,
  StatusItemIconAlign,
  StatusItemsPosition,
  StatusItemsProps,
} from './';
import { Icon } from '../Icon';
import { Popup } from '../Popup';
import { Tooltip } from '../Tooltip';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { ConditionalWrapper, mergeClasses } from '../../shared/utilities';

import styles from './avatar.module.scss';

export const AVATAR_THEME_SET = [
  styles.red,
  styles.redOrange,
  styles.orange,
  styles.yellow,
  styles.yellowGreen,
  styles.green,
  styles.blueGreen,
  styles.blue,
  styles.blueViolet,
  styles.violet,
  styles.violetRed,
  styles.grey,
];

export const getStatusItemSizeAndPadding = (
  avatarSize: number
): [number, number] => {
  // Returns: [status item size, status item padding]
  const statusItemSize: number = (avatarSize * 16) / 100;
  return [statusItemSize, (statusItemSize * 6) / 16];
};

// 0.06 factor is chosen based on design
const StatusItemWrapperPaddingFactor: number = 0.06;
const DefaultStatusItemMaxTextLength: number = 3;
const MinStatusItemFontSize: number = 12;
const StatusItemFontDiff: string = '2px';

const StatusItemOutlineDefaults: React.CSSProperties = {
  outlineColor: 'var(--grey-color)',
  outlineOffset: '0px',
  outlineStyle: 'solid',
  outlineWidth: '2px',
};

const AvatarOutlineDefaults: React.CSSProperties = {
  outlineColor: 'var(--green-tertiary-color)',
  outlineOffset: '2px',
  outlineStyle: 'solid',
  outlineWidth: '4px',
};

const statusItemPositionRtlMap: {
  [key in StatusItemsPosition]: StatusItemsPosition;
} = {
  [StatusItemsPosition.Top]: StatusItemsPosition.Top,
  [StatusItemsPosition.Bottom]: StatusItemsPosition.Bottom,
  [StatusItemsPosition.Left]: StatusItemsPosition.Right,
  [StatusItemsPosition.Right]: StatusItemsPosition.Left,
  [StatusItemsPosition.TopRight]: StatusItemsPosition.TopLeft,
  [StatusItemsPosition.TopLeft]: StatusItemsPosition.TopRight,
  [StatusItemsPosition.BottomRight]: StatusItemsPosition.BottomLeft,
  [StatusItemsPosition.BottomLeft]: StatusItemsPosition.BottomRight,
};

const AvatarStatusItems: FC<BaseAvatarProps> = React.forwardRef(
  ({ outline, size, statusItems }, ref: Ref<HTMLDivElement>) => {
    const statusItemsRef = useRef<{
      [key in StatusItemsPosition]?: HTMLSpanElement;
    }>({});
    const [showStatusItemsText, setShowStatusItemsText] = useState<{
      [key in StatusItemsPosition]?: boolean;
    }>({});
    const htmlDir: string = useCanvasDirection();

    const updateStatusItemTextVisibility = (position: StatusItemsPosition) => {
      const value = statusItemsRef.current[position];
      if (value === undefined) {
        return;
      }
      const styles = getComputedStyle(value); // getComputedStyle always outputs a pixel value
      // We do slice to remove "px"
      setShowStatusItemsText((prevState) => ({
        ...prevState,
        [position]:
          parseInt(styles.fontSize.slice(0, -2)) >= MinStatusItemFontSize,
      }));
    };

    useEffect(() => {
      updateStatusItemTextVisibility(StatusItemsPosition.Top);
      updateStatusItemTextVisibility(StatusItemsPosition.Bottom);
      updateStatusItemTextVisibility(StatusItemsPosition.Left);
      updateStatusItemTextVisibility(StatusItemsPosition.Right);
      updateStatusItemTextVisibility(StatusItemsPosition.TopRight);
      updateStatusItemTextVisibility(StatusItemsPosition.TopLeft);
      updateStatusItemTextVisibility(StatusItemsPosition.BottomRight);
      updateStatusItemTextVisibility(StatusItemsPosition.BottomLeft);
    }, [statusItemsRef.current]);

    const getStatusItemPositionStyle = (
      itemPos: StatusItemsPosition
    ): React.CSSProperties => {
      const outlineWidth: string = outline?.outlineWidth ?? '0px'; // Avatar outline width
      const outlineOffset: string = outline?.outlineOffset ?? '0px'; // Avatar outline offset
      const radius: string = `calc((${size} + ${outlineWidth} + ${outlineOffset}) / 2)`;

      switch (htmlDir === 'rtl' ? statusItemPositionRtlMap[itemPos] : itemPos) {
        case StatusItemsPosition.TopRight:
          return {
            transform: `rotate(${-45}deg) translate(${radius}) rotate(${45}deg)`,
          };
        case StatusItemsPosition.TopLeft:
          return {
            transform: `rotate(${-135}deg) translate(${radius}) rotate(${135}deg)`,
          };
        case StatusItemsPosition.BottomRight:
          return {
            transform: `rotate(${45}deg) translate(${radius}) rotate(${-45}deg)`,
          };
        case StatusItemsPosition.BottomLeft:
          return {
            transform: `rotate(${135}deg) translate(${radius}) rotate(${-135}deg)`,
          };
        case StatusItemsPosition.Left:
          return {
            transform: `rotate(${180}deg) translate(${radius}) rotate(${-180}deg)`,
          };
        case StatusItemsPosition.Right:
          return { transform: `translate(${radius})` };
        case StatusItemsPosition.Top:
          return {
            transform: `rotate(${-90}deg) translate(${radius}) rotate(${90}deg)`,
          };
        case StatusItemsPosition.Bottom:
        default:
          return {
            transform: `rotate(${90}deg) translate(${radius}) rotate(${-90}deg)`,
          };
      }
    };

    return (
      <>
        {Object.keys(statusItems).map((position: StatusItemsPosition) => {
          const statusItemProps: StatusItemsProps = statusItems[position];
          const alignIcon: StatusItemIconAlign =
            statusItemProps.alignIcon ?? StatusItemIconAlign.Right;
          const showStatusItemText: boolean =
            statusItemProps.text &&
            statusItemProps.text.length <=
              (statusItemProps.textMaxLength ?? DefaultStatusItemMaxTextLength);
          const wrapperPadding: string | number =
            statusItemProps?.wrapperStyle?.padding ??
            `(${size} * ${StatusItemWrapperPaddingFactor})`;
          const statusItemTextClasses = mergeClasses([
            styles.avatarStatusItemText,
            { [styles.avatarStatusItemTextRtl]: htmlDir === 'rtl' },
            {
              [styles.textMarginRight]: alignIcon == StatusItemIconAlign.Right,
            },
            { [styles.textMarginLeft]: alignIcon == StatusItemIconAlign.Left },
          ]);
          const statusItemIconElement = (
            <Icon
              {...statusItemProps}
              classNames={styles.avatarStatusItemIcon}
            />
          );
          return (
            <div
              key={position}
              ref={ref}
              style={{
                background:
                  statusItemProps.backgroundColor ??
                  'var(--avatar-status-item-background)',
                padding: `calc(${wrapperPadding})`,
                ...getStatusItemPositionStyle(position),
                ...(statusItemProps.outline
                  ? {
                      outlineColor:
                        statusItemProps.outline?.outlineColor ??
                        StatusItemOutlineDefaults.outlineColor,
                      outlineOffset:
                        statusItemProps.outline?.outlineOffset ??
                        StatusItemOutlineDefaults.outlineOffset,
                      outlineStyle:
                        statusItemProps.outline?.outlineStyle ??
                        StatusItemOutlineDefaults.outlineStyle,
                      outlineWidth:
                        statusItemProps.outline?.outlineWidth ??
                        StatusItemOutlineDefaults.outlineWidth,
                    }
                  : {}),
                ...(statusItemProps.wrapperStyle ?? {}),
              }}
              className={`${styles.avatarStatusItem} ${
                statusItemProps.wrapperClassName ?? ''
              } ${statusItemProps.onClick ? styles.clickable : ''}`}
              {...(statusItemProps.onClick
                ? {
                    onClick: statusItemProps.onClick,
                    role: 'button',
                  }
                : {})}
              {...(statusItemProps.ariaLabel
                ? { 'aria-label': statusItemProps.ariaLabel }
                : {})}
            >
              {alignIcon == StatusItemIconAlign.Left && statusItemIconElement}
              {showStatusItemText && (showStatusItemsText[position] ?? true) && (
                <span
                  ref={(el) => (statusItemsRef.current[position] = el)}
                  style={{
                    color: statusItemProps.color,
                    fontSize: `calc(${statusItemProps.size} + ${StatusItemFontDiff})`,
                    lineHeight: statusItemProps.size,
                  }}
                  className={statusItemTextClasses}
                >
                  {statusItemProps.text}
                </span>
              )}
              {alignIcon == StatusItemIconAlign.Right && statusItemIconElement}
            </div>
          );
        })}
      </>
    );
  }
);

const AvatarFallback: FC<AvatarFallbackProps> = React.forwardRef(
  (
    {
      children,
      classNames,
      hashingFunction,
      onClick,
      onKeyDown,
      onMouseEnter,
      onMouseLeave,
      randomiseTheme,
      style,
      theme,
    },
    ref: Ref<HTMLDivElement>
  ) => {
    const colorSetIndex: number = useMemo(() => {
      if (randomiseTheme) {
        return Math.floor(Math.random() * 100) % AVATAR_THEME_SET.length;
      }
      if (hashingFunction) {
        return Math.floor(hashingFunction()) % AVATAR_THEME_SET.length;
      }
      return -1;
    }, []);

    const avatarClasses: string = mergeClasses([
      styles.wrapperStyle,
      classNames,
      { [styles.red]: theme === 'red' },
      { [styles.redOrange]: theme === 'redOrange' },
      { [styles.orange]: theme === 'orange' },
      { [styles.yellow]: theme === 'yellow' },
      { [styles.yellowGreen]: theme === 'yellowGreen' },
      { [styles.green]: theme === 'green' },
      { [styles.blueGreen]: theme === 'blueGreen' },
      { [styles.blue]: theme === 'blue' },
      { [styles.blueViolet]: theme === 'blueViolet' },
      { [styles.violet]: theme === 'violet' },
      { [styles.violetRed]: theme === 'violetRed' },
      { [styles.grey]: theme === 'grey' },
      AVATAR_THEME_SET?.[colorSetIndex],
    ]);

    return (
      <div
        ref={ref}
        className={avatarClasses}
        onClick={onClick}
        onKeyDown={onKeyDown}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        style={style}
        tabIndex={0}
      >
        {children}
      </div>
    );
  }
);

const AvatarIcon: FC<AvatarIconProps> = React.forwardRef(
  (
    {
      children,
      classNames,
      fontSize,
      iconProps,
      onClick,
      onKeyDown,
      onMouseEnter,
      onMouseLeave,
      style,
    },
    ref: Ref<HTMLDivElement>
  ) => {
    const wrapperClasses: string = mergeClasses([
      styles.wrapperStyle,
      classNames,
    ]);

    return (
      <div
        ref={ref}
        className={wrapperClasses}
        onClick={onClick}
        onKeyDown={onKeyDown}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        style={style}
        tabIndex={0}
      >
        <Icon size={fontSize} {...iconProps} />
        {children}
      </div>
    );
  }
);

export const Avatar: FC<AvatarProps> = React.forwardRef(
  (
    {
      alt,
      children,
      classNames,
      fontSize = '18px',
      hashingFunction,
      key,
      iconProps,
      onClick,
      onKeyDown,
      onMouseEnter,
      onMouseLeave,
      outline,
      popupProps = undefined,
      randomiseTheme,
      size = '32px',
      src,
      statusItems = {},
      style = {},
      theme,
      tooltipProps = undefined,
      type = 'square',
    },
    ref: Ref<HTMLDivElement>
  ) => {
    const htmlDir: string = useCanvasDirection();

    const [popupTriggerSize, setPopupTriggerSize] = useState<number>(
      parseInt(size, 10)
    );
    const [popupVisible, setPopupVisibility] = useState<boolean>(false);

    let calculatedOutline: AvatarOutlineProps = undefined;
    if (outline !== undefined) {
      calculatedOutline = {
        outlineColor:
          outline?.outlineColor ?? AvatarOutlineDefaults.outlineColor,
        outlineOffset:
          outline?.outlineOffset ??
          AvatarOutlineDefaults.outlineOffset.toString(),
        outlineStyle:
          outline?.outlineStyle ?? AvatarOutlineDefaults.outlineStyle,
        outlineWidth:
          outline?.outlineWidth ??
          AvatarOutlineDefaults.outlineWidth.toString(),
      };
    }

    useEffect(() => {
      setPopupTriggerSize(parseInt(size, 10));
    }, [size]);

    const popupClassNames: string = mergeClasses([
      { [styles.avatarPopup]: popupProps && !!popupVisible },
      { [styles.avatarPopupVisible]: popupProps && !!popupVisible },
      { [styles.avatarPopupHidden]: popupProps && !popupVisible },
      { [styles.round]: type === 'round' },
    ]);

    const imageClassNames: string = mergeClasses([
      styles.avatar,
      styles.imageStyle,
      popupClassNames,
    ]);

    const wrapperContainerClassNames: string = mergeClasses([
      styles.avatarImgWrapper,
      popupClassNames,
      classNames,
    ]);

    const wrapperContainerStyle: React.CSSProperties = {
      width: size,
      height: size,
      minWidth: size,
      minHeight: size,
      fontSize: fontSize,
      ...style,
      ...(Object.keys(statusItems).length > 0 ? { position: 'relative' } : {}),
    };

    const getPopup = (children: React.ReactNode): JSX.Element => (
      <Popup
        closeOnPopupClick={false}
        key={`${key}-popup`}
        minHeight={Math.floor(popupTriggerSize + popupTriggerSize / 1.5)}
        offset={-Math.floor(popupTriggerSize / 2)}
        onVisibleChange={(isVisible) => setPopupVisibility(isVisible)}
        placement="bottom-start"
        popupStyle={{
          margin:
            htmlDir === 'rtl'
              ? `0 ${Math.floor(popupTriggerSize / 4)}px`
              : `0 -${Math.floor(popupTriggerSize / 4)}px`,
          padding: `${Math.floor(popupTriggerSize / 3)}px`,
          paddingTop: `${Math.floor(popupTriggerSize / 1.5)}px`,
          borderRadius: `${Math.floor(popupTriggerSize / 4)}px`,
        }}
        portal
        tabIndex={-1} // Prevent focus on the reference wrapper, defer to Avatar
        triggerAbove
        visibleArrow={false}
        width={Math.floor(popupTriggerSize * 4)}
        {...popupProps}
      >
        {children}
      </Popup>
    );

    if (src) {
      return (
        <ConditionalWrapper
          condition={tooltipProps !== undefined || popupProps !== undefined}
          wrapper={(children: React.ReactNode): JSX.Element => (
            <>
              {tooltipProps && popupProps === undefined && (
                <Tooltip
                  classNames={styles.avatarTooltip}
                  key={`${key}-tooltip`}
                  tabIndex={-1}
                  {...tooltipProps}
                >
                  {children}
                </Tooltip>
              )}
              {popupProps && getPopup(children)}
            </>
          )}
        >
          <div
            className={wrapperContainerClassNames}
            ref={ref}
            style={wrapperContainerStyle}
          >
            <img
              alt={alt}
              className={imageClassNames}
              height={size}
              onClick={onClick}
              onKeyDown={onKeyDown}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
              src={src}
              style={calculatedOutline}
              tabIndex={0}
              width={size}
            />
            <AvatarStatusItems
              outline={calculatedOutline}
              size={size}
              statusItems={statusItems}
            />
          </div>
        </ConditionalWrapper>
      );
    }

    const wrapperClassNames: string = mergeClasses([
      imageClassNames,
      classNames,
    ]);

    const wrapperChildClassNames: string = mergeClasses([
      popupClassNames,
      classNames,
    ]);

    if (iconProps) {
      return (
        <ConditionalWrapper
          condition={tooltipProps !== undefined || popupProps !== undefined}
          wrapper={(children: React.ReactNode): JSX.Element => (
            <>
              {tooltipProps && popupProps === undefined && (
                <Tooltip
                  classNames={styles.avatarTooltip}
                  key={`${key}-tooltip`}
                  tabIndex={-1}
                  {...tooltipProps}
                >
                  <div>{children}</div>
                </Tooltip>
              )}
              {popupProps &&
                getPopup(
                  <div className={wrapperChildClassNames}>{children}</div>
                )}
            </>
          )}
        >
          <AvatarIcon
            classNames={wrapperClassNames}
            fontSize={fontSize}
            iconProps={iconProps}
            onClick={onClick}
            onKeyDown={onKeyDown}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            ref={ref}
            style={{ ...wrapperContainerStyle, ...(calculatedOutline ?? {}) }}
          >
            <AvatarStatusItems
              outline={calculatedOutline}
              size={size}
              statusItems={statusItems}
            />
          </AvatarIcon>
        </ConditionalWrapper>
      );
    }

    return (
      <ConditionalWrapper
        condition={tooltipProps !== undefined || popupProps !== undefined}
        wrapper={(children: React.ReactNode): JSX.Element => (
          <>
            {tooltipProps && popupProps === undefined && (
              <Tooltip
                classNames={styles.avatarTooltip}
                key={`${key}-tooltip`}
                tabIndex={-1}
                {...tooltipProps}
              >
                <div>{children}</div>
              </Tooltip>
            )}
            {popupProps &&
              getPopup(
                <div className={wrapperChildClassNames}>{children}</div>
              )}
          </>
        )}
      >
        <AvatarFallback
          classNames={wrapperClassNames}
          hashingFunction={hashingFunction}
          onClick={onClick}
          onKeyDown={onKeyDown}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          randomiseTheme={randomiseTheme}
          ref={ref}
          style={{ ...wrapperContainerStyle, ...(calculatedOutline ?? {}) }}
          theme={theme}
        >
          {children}
          <AvatarStatusItems
            outline={calculatedOutline}
            size={size}
            statusItems={statusItems}
          />
        </AvatarFallback>
      </ConditionalWrapper>
    );
  }
);
