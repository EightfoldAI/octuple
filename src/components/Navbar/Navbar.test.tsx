import React, { createContext, useState } from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Avatar } from '../Avatar';
import { ConfigProvider, useConfig } from '../ConfigProvider';
import { IConfigContext } from '../ConfigProvider/ConfigProvider.types';
import { Dropdown } from '../Dropdown';
import { Icon, IconName } from '../Icon';
import { Link } from '../Link';
import { List } from '../List';
import { Navbar, NavbarContent } from './';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

const ConfigContext: React.Context<Partial<IConfigContext>> = createContext<
  Partial<IConfigContext>
>({});

const sampleList: string[] = [1, 2, 3, 4, 5].map((i) => `User ${i}`);

const Overlay = () => {
  return (
    <List
      items={sampleList}
      layout="vertical"
      renderItem={(item: string) => <div style={{ margin: '8px' }}>{item}</div>}
    />
  );
};

const ProfileDropdown = () => {
  const [visible, setVisibility] = useState(false);
  return (
    <Dropdown
      onVisibleChange={(isVisible) => setVisibility(isVisible)}
      overlay={<Overlay />}
      dropdownStyle={{ color: '#212121' }}
    >
      <div style={{ display: 'flex' }}>
        <Avatar
          src="https://images.pexels.com/photos/771742/pexels-photo-771742.jpeg"
          alt="random profile image"
          type="round"
        />
        <Icon path={IconName.mdiChevronDown} rotate={visible ? 180 : 0} />
      </div>
    </Dropdown>
  );
};

describe('Navbar', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Renders without crashing', () => {
    const { container, getByTestId } = render(
      <ConfigProvider
        themeOptions={{
          name: 'blue',
        }}
      >
        <div style={{ height: '250px' }}>
          <Navbar data-testid="test-navbar">
            <NavbarContent>
              <Link
                href="https://eightfold.ai"
                target="_blank"
                variant="default"
                style={{ padding: '8px 20px', color: 'inherit' }}
              >
                Eightfold
              </Link>
            </NavbarContent>
            <NavbarContent>
              <ProfileDropdown />
            </NavbarContent>
          </Navbar>
        </div>
      </ConfigProvider>
    );
    const navbar = getByTestId('test-navbar');
    expect(() => container).not.toThrowError();
    expect(navbar).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('Navbar theme if props are provided', () => {
    const { result } = renderHook(() => useConfig(), {
      wrapper: ({ children }) => (
        <ConfigProvider
          themeOptions={{
            name: 'nav',
            customTheme: {
              navbarTheme: {
                background: '#3E5334',
                textColor: '#EAD7A4',
                textHoverColor: '#3E5334',
                textHoverBackground: '#8FA394',
              },
            },
          }}
        >
          {children}
        </ConfigProvider>
      ),
    });
    expect(result.current.themeOptions).toEqual({
      name: 'nav',
      customTheme: {
        navbarTheme: {
          background: '#3E5334',
          textColor: '#EAD7A4',
          textHoverColor: '#3E5334',
          textHoverBackground: '#8FA394',
        },
      },
    });
  });

  test('Navbar matches custom theme', () => {
    const { container } = render(
      <ConfigProvider
        themeOptions={{
          name: 'nav',
          customTheme: {
            navbarTheme: {
              background: '#3E5334',
              textColor: '#EAD7A4',
              textHoverColor: '#3E5334',
              textHoverBackground: '#8FA394',
            },
          },
        }}
      >
        <Navbar data-testid="test-navbar">
          <NavbarContent>
            <Link
              href="https://eightfold.ai"
              target="_blank"
              variant="primary"
              style={{
                backgroundColor: 'transparent',
                color: 'var(--navbar-text-color)',
                padding: '8px 20px',
              }}
              data-testid="test-navbar-link"
            >
              Eightfold
            </Link>
          </NavbarContent>
          <NavbarContent>
            <ProfileDropdown />
          </NavbarContent>
        </Navbar>
      </ConfigProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
