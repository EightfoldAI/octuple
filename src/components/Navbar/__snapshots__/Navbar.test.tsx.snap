// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Navbar matches custom theme 1`] = `
<div>
  <div
    class="navbar-container"
    data-testid="test-navbar"
  >
    <div
      class="navbar-content"
    >
      <a
        aria-disabled="false"
        class="link-style full-width primary"
        data-testid="test-navbar-link"
        href="https://eightfold.ai"
        role="link"
        style="background-color: transparent; padding: 8px 20px;"
        target="_blank"
      >
        Eightfold
      </a>
    </div>
    <div
      class="navbar-content"
    >
      <div
        class="main-wrapper"
      >
        <div
          aria-controls="dropdown-"
          aria-expanded="false"
          aria-haspopup="true"
          class="reference-wrapper"
          id="dropdown-reference"
          role="button"
          style="display: flex;"
          tabindex="0"
        >
          <div
            class="avatar-img-wrapper round"
            style="width: 32px; height: 32px; min-width: 32px; min-height: 32px; font-size: 18px;"
          >
            <img
              alt="random profile image"
              class="avatar image-style round"
              height="32px"
              src="https://images.pexels.com/photos/771742/pexels-photo-771742.jpeg"
              tabindex="0"
              width="32px"
            />
          </div>
          <span
            aria-hidden="false"
            class="icon-wrapper"
            role="presentation"
          >
            <svg
              role="presentation"
              style="width: 20px; height: 20px;"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
                style="fill: currentColor;"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Navbar Renders without crashing 1`] = `
<div>
  <div
    style="height: 250px;"
  >
    <div
      class="navbar-container"
      data-testid="test-navbar"
    >
      <div
        class="navbar-content"
      >
        <a
          aria-disabled="false"
          class="link-style full-width"
          href="https://eightfold.ai"
          role="link"
          style="padding: 8px 20px;"
          target="_blank"
        >
          Eightfold
        </a>
      </div>
      <div
        class="navbar-content"
      >
        <div
          class="main-wrapper"
        >
          <div
            aria-controls="dropdown-"
            aria-expanded="false"
            aria-haspopup="true"
            class="reference-wrapper"
            id="dropdown-reference"
            role="button"
            style="display: flex;"
            tabindex="0"
          >
            <div
              class="avatar-img-wrapper round"
              style="width: 32px; height: 32px; min-width: 32px; min-height: 32px; font-size: 18px;"
            >
              <img
                alt="random profile image"
                class="avatar image-style round"
                height="32px"
                src="https://images.pexels.com/photos/771742/pexels-photo-771742.jpeg"
                tabindex="0"
                width="32px"
              />
            </div>
            <span
              aria-hidden="false"
              class="icon-wrapper"
              role="presentation"
            >
              <svg
                role="presentation"
                style="width: 20px; height: 20px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
                  style="fill: currentColor;"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
